cpl/module_block_data.F90

```diff
@@ -0,0 +1,483 @@
+
+  end subroutine block_copy_3d_to_3d_r8
+
+  subroutine block_array_copy_3d_to_3d_r8(destin_ptr, source_arr, block, block_index, scale_factor)
+
+    ! -- arguments
+    real(ESMF_KIND_R8),        pointer    :: destin_ptr(:,:,:)
+    real,                      intent(in) :: source_arr(:,:,:)
+    type (block_control_type), intent(in) :: block
+    integer,                   intent(in) :: block_index
+    real, optional,            intent(in) :: scale_factor
+
+    ! -- local variables
+    integer         :: i, ib, ix, j, jb, k
+    real(kind_phys) :: factor
+
+    ! -- begin
+    if (associated(destin_ptr)) then
+      factor = 1._kind_phys
+      if (present(scale_factor)) factor = scale_factor
+      do k = 1, size(source_arr, dim=3)
+        do ix = 1, block%blksz(block_index)
+          ib = block%index(block_index)%ii(ix)
+          jb = block%index(block_index)%jj(ix)
+          i = ib - block%isc + 1
+          j = jb - block%jsc + 1
+          destin_ptr(i,j,k) = factor * source_arr(ib,jb,k)
+        enddo
+      enddo
+    end if
+
+  end subroutine block_array_copy_3d_to_3d_r8
+
+  ! -- copy: 3D slice to 3D
+
+  subroutine block_copy_3dslice_to_3d_r8(destin_ptr, source_ptr, slice, block, block_index, scale_factor)
+
+    ! -- arguments
+    real(ESMF_KIND_R8),        pointer    :: destin_ptr(:,:,:)
+    real(kind_phys),           pointer    :: source_ptr(:,:,:,:)
+    integer,                   intent(in) :: slice
+    type (block_control_type), intent(in) :: block
+    integer,                   intent(in) :: block_index
+    real(kind_phys), optional, intent(in) :: scale_factor
+
+    ! -- local variables
+    integer         :: i, ib, ix, j, jb, k
+    real(kind_phys) :: factor
+
+    ! -- begin
+    if (associated(destin_ptr) .and. associated(source_ptr)) then
+      if (slice > 0 .and. slice <= size(source_ptr, dim=4)) then
+        factor = 1._kind_phys
+        if (present(scale_factor)) factor = scale_factor
+        do k = 1, size(source_ptr, dim=3)
+          do ix = 1, block%blksz(block_index)
+            ib = block%index(block_index)%ii(ix)
+            jb = block%index(block_index)%jj(ix)
+            i = ib - block%isc + 1
+            j = jb - block%jsc + 1
+            destin_ptr(i,j,k) = factor * source_ptr(ib,jb,k,slice)
+          enddo
+        enddo
+      end if
+    end if
+
+  end subroutine block_copy_3dslice_to_3d_r8
+
+  subroutine block_array_copy_3dslice_to_3d_r8(destin_ptr, source_arr, slice, block, block_index, scale_factor)
+
+    ! -- arguments
+    real(ESMF_KIND_R8),        pointer    :: destin_ptr(:,:,:)
+    real,                      intent(in) :: source_arr(:,:,:,:)
+    integer,                   intent(in) :: slice
+    type (block_control_type), intent(in) :: block
+    integer,                   intent(in) :: block_index
+    real, optional,            intent(in) :: scale_factor
+
+    ! -- local variables
+    integer         :: i, ib, ix, j, jb, k
+    real(kind_phys) :: factor
```

Comment:
Why is `factor`'s type `kind_phys`, when both `scale_factor` and `source_arr` are default real?

Reply:
Good point! Will fix it.

Reply:
Commit 2adafb484cfadfadf6fdcfc2cd4624174b713253 fixed this issue.